File: prettier-plugin-csharpier/src/Printer/Types/CatchClause.ts
import { PrintMethod } from "../PrintMethod";
import { SyntaxToken, printPathSyntaxToken, SyntaxTreeNode, HasIdentifier, printIdentifier } from "../SyntaxTreeNode";
import { concat, group, hardline, indent, join, softline, line, doubleHardline } from "../Builders";
import { BlockNode, printBlock } from "./Block";

export interface CatchClauseNode extends SyntaxTreeNode<"CatchClause"> {
    catchKeyword?: SyntaxToken;
    declaration?: CatchDeclarationNode;
    filter?: CatchFilterClauseNode;
    block?: BlockNode;
}

interface CatchDeclarationNode extends SyntaxTreeNode<"CatchDeclaration"> {
    openParenToken?: SyntaxToken;
    type?: SyntaxTreeNode;
    identifier: SyntaxToken;
    closeParenToken?: SyntaxToken;
}

interface CatchFilterClauseNode extends SyntaxTreeNode<"CatchFilterClause"> {
    whenKeyword?: SyntaxToken;
    openParenToken?: SyntaxToken;
    filterExpression?: SyntaxTreeNode;
    closeParenToken?: SyntaxToken;
}

export const printCatchClause: PrintMethod<CatchClauseNode> = (path, options, print) => {
    const node = path.getValue();
    return concat([
        "catch",
        node.declaration
            ? path.call(catchDeclarationPath => {
                  const catchDeclarationNode = catchDeclarationPath.getValue();
                  return concat([
                      " (",
                      catchDeclarationPath.call(print, "type"),
                      catchDeclarationNode.identifier ? " " : "",
                      printIdentifier(catchDeclarationNode),
                      ")"
                  ]);
              }, "declaration")
            : "",
        node.filter
            ? path.call(catchFilterClausePath => {
                  return concat([" when (", catchFilterClausePath.call(print, "filterExpression"), ")"]);
              }, "filter")
            : "",
        path.call(o => printBlock(o, options, print), "block")
    ]);
};

File: Src/CSharpier/SyntaxPrinter/SyntaxNodePrinters/CatchClause.cs
namespace CSharpier.SyntaxPrinter.SyntaxNodePrinters;

internal static class CatchClause
{
    public static Doc Print(CatchClauseSyntax node, FormattingContext context)
    {
        return Doc.Concat(
            Token.Print(node.CatchKeyword, context),
            Doc.Group(
                node.Declaration != null
                    ? Doc.Concat(
                        " ",
                        Token.Print(node.Declaration.OpenParenToken, context),
                        Node.Print(node.Declaration.Type, context),
                        node.Declaration.Identifier.RawSyntaxKind() != SyntaxKind.None
                            ? " "
                            : Doc.Null,
                        Token.Print(node.Declaration.Identifier, context),
                        Token.Print(node.Declaration.CloseParenToken, context)
                    )
                    : Doc.Null,
                node.Filter != null
                    ? Doc.Indent(
                        Doc.Line,
                        Token.PrintWithSuffix(node.Filter.WhenKeyword, " ", context),
                        Token.Print(node.Filter.OpenParenToken, context),
                        Doc.Group(
                            Doc.Indent(Node.Print(node.Filter.FilterExpression, context)),
                            Doc.SoftLine
                        ),
                        Token.Print(node.Filter.CloseParenToken, context)
                    )
                    : Doc.Null
            ),
            Block.Print(node.Block, context)
        );
    }
}
